feat(treesitter): highlighting for core languages, enabled for Lua

vigoux opened the pull request:
Part of the plan for 0.6 is to support C, Lua and VimL out of the box with tree-sitter.

In order to move forward with the plan, add lua into the parsers shiped by default.

vigoux committed:
feat(treesitter): bundle Lua parser and queries

parser from https://github.com/MunifTanjim/tree-sitter-lua
queries from nvim-treesitter

stsewd commented:
Probably we should have an option to be able to disable/override them too, we already hit a problem with the C parser when an update was pushed to the nvim-treesitter repo and the queries were still read from the neovim repo.

vigoux commented:
Indeed that's a problem, I'll try to see what I can come up with

justinmk commented on runtime/queries/lua/highlights.scm:
```diff
@@ -0,0 +1,159 @@
+;;; Highlighting for lua
```
what are the tradeoffs if we put these `*.scm` files in `runtime/syntax/` instead of a new `runtime/queries` location? (Other than the fact that Nvim 0.5 shipped with `queries/`)

theHamsta commented:
I don't think this is a good idea:
- the parsers are rather big (0.1-1MB for most parsers) so it won't be possible to include parsers for all languages. Why use a different highlighting system for some languages and for others not? An average programmer won't have any immediate advantage from using Tree-Sitter, especially for niche languages like C and viml. 
- bonus surprise effect: with only  the C parser installed you will have different highlighting in header and source files as vim will default-guess headers as C++ files
- built-in runtime files have the disadvantage that they age quite quickly. This can be easily fixed when a plug-in provides an updated version of that specific runtime file. Parser update is more complicated. Parsers have to be updated in conjunction with all dependent `*.scm` files. It might be more complicated to support multiple old released version of Neovim when parsers are bundled with a release instead of providing a method to install the newest release of the parsers. Can a third-party plug-in easily provide an update for a parser while other plug-in a still target the versions of built-in parsers with their scm files? How will version updates be synchronized between between built-in parsers and scm-providers? Can plug-ins offer multiple runtime file versions to be compatible with nvim 0.6 and 0.6.1and 0.7?

A neutral point:
- the default experience is currently shared and maintained with the upstream Vim project. There is nothing wrong with syntax files to provide a basic default experience. Having a different system might be distinctive but might also involve additional maintenance.

My suggested solutions:
- treat parsers like spell files: Neovim just provides an method to download the newest or a compatible release of a parser. A method to release and install in the users favorite form pre-compiled/source could be realized within the 1.0 milestone of Tree-Sitter project and would be externally maintained. Users only have installed what they requested.
- parsers are installed together with `*.scm`as part of language support plug-in (same source for SCM and parser)

justinmk commented:
> * treat parsers like spell files: Neovim just provides an method to download the newest or a compatible release of a parser. A method to release and install in the users favorite form pre-compiled/source could be realized within the 1.0 milestone of Tree-Sitter project and would be externally maintained. Users only have installed what they requested.

This is mostly what our plan has been, though with some aggressive nudging towards making it the default experience.

I don't see much risk in inlining the C and Lua stuff while we figure out a longer-term approach. But I think we're generally aligned with your concerns.

clason commented:
Revisiting this for the 0.8 milestone. I disagree with @theHamsta on some points (although the technical points are of course uncontested):

* Lua and Vimscript _are_ special, since these are "our" languages, and using them is intrinsically tied to using Neovim. Hence we should do everything we can to improve that experience.
* Lua and Vimscript are also special because they make heavy use of injections: Lua into Vimscript, Vimscript into Lua, and C into Lua (via FFI). This is exactly what tree-sitter is good at, and better than regex syntax.

So I think it makes sense to at least bundle Lua, Vim, and C parsers and queries (with the C parser only used for injections and not standalone files, for the reason mentioned above), making them opt-in for 0.8 (and possibly opt-out for 0.9). Comment and query parsers are also possible options (especially the former, as that is one people tend to miss when installing manually).

Of course, clearly more work (and thought) is needed to get the different parts (integration, parsers, queries) in the necessary shape, but we have to start working towards this at _some_ point -- and I believe that now is a good point.

clason commented:
One thing this needs is better control over loading/overriding parsers and (to a lesser extent) queries. We should decide on and document some standard locations in shipped and user runtime and push for plugins to respect those.

clason commented:
Suggestion: bundle parser and queries together: `ts/<lang>/{parser,queries}/*` instead of `parsers/<lang>.so` and `queries/<lang>/*.scm` -- makes it easier to load matching parser and queries if multiple exist in runtime.

dundargoc commented on cmake.deps/cmake/BuildTreesitterParsers.cmake:
```diff
@@ -1,29 +1,32 @@
+      ${DEPS_BUILD_DIR}/src/${NAME}/CMakeLists.txt
+  CMAKE_ARGS
+    -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
+    -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
```
This variable is unused. Delete.

clason replied:
Is that still the case when building a  `scanner.cc`?

theHamsta replied:
I don't think `CMAKE_ARGS` are necessary since the defaults are not changed

clason replied:
nor is the patch command?

dundargoc replied:
> Is that still the case when building a scanner.cc?

Unsure. I will test.

dundargoc replied:
The Treesittercxx doesn't work properly. Let's remove `CMAKE_CXX_COMPILER` for now since it gives a warning and deal with it later if we need a C++ parser.

clason replied:
Anything else I should remove from that file? Everything but `-DPARSERLANG` should be default.

dundargoc replied:
> nor is the patch command?

Nah, the "patching" is necessary (technically not a patch, but it's the most convenient way to get cmake to work.)

You can remove all arguments from `CMAKE_ARGS` (including the toolchain comment) except for `-DPARSERLANG` and 
`-DCMAKE_INSTALL_PREFIX`. Default for `CMAKE_INSTALL_PREFIX` is `/usr/local`.

dundargoc commented on cmake.deps/cmake/BuildTreesitterParsers.cmake:
```diff
@@ -1,29 +1,32 @@
+    -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
+    -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
+    -DCMAKE_GENERATOR=${CMAKE_GENERATOR}
+    -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
```
This needs to be changed to a special variable, otherwise a warning will appear on Windows.

```suggestion
    ${BUILD_TYPE_STRING}
```

dundargoc commented on cmake.deps/cmake/TreesitterParserCMakeListsCXX.txt:
```diff
@@ -0,0 +1,19 @@
+cmake_minimum_required(VERSION 3.10)
```
This entire file doesn't seem to be even used. Is there a reason for this?

clason replied:
Not yet; but if we include other parsers that use a `scanner.cc` instead of a `scanner.c`, we will?

clason replied:
Should be reflected in the name, though -- maybe `TreeSitterParserCMakeListsScannerC` and (possibly later) `TreeSitterParserCMakeListsScannerCXX`?

clason replied:
@dundargoc Is there any reasonable way to combine these three files into a single one? We have three situations, depending on the specific parser:
1. use a C compiler to compile `src/parser.c`
2. use a C compiler to compile `src/parser.c` and `src/scanner.c`
3. use a C++ compiler to compile `src/parser.c` and `src/scanner.cc`

(Current plan -- C, Lua, Vimscript -- only needs options 1 and 2.)

dundargoc replied:
1 and 2 can be combined for sure; my [comment](https://github.com/neovim/neovim/pull/15391#issuecomment-1219515823) outlines how.

The third one should be doable but I'd need to test on a c++ parser to be sure.

clason replied:
I'm a bit wary that some parsers include extra files in `src`, but we can cross that bridge when we come to it.

https://github.com/tree-sitter/tree-sitter-python uses a `scanner.cc`, for example.

dundargoc commented on cmake.deps/cmake/BuildTreesitterParsers.cmake:
```diff
@@ -1,29 +1,32 @@
+    # Pass toolchain
+    -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TOOLCHAIN_FILE}
+    -DPARSERLANG=${LANG}
+  BUILD_COMMAND ${CMAKE_COMMAND} --build . --config ${CMAKE_BUILD_TYPE}
```
You can delete the line with `BUILD_COMMAND` and `INSTALL_COMMAND`, cmake will do those automatically.

dundargoc commented:
I don't think we need `TreesitterParserCMakeListsExternal.txt`, I think a single `TreesitterParserCMakeLists.txt` should be enough for both parsers.

If we change

```
 add_library(parser
   MODULE
   src/parser.c
 )
```

to

```
file(GLOB source_files src/*.c)

add_library(parser
  MODULE
  ${source_files}
)
```

then it works for both parsers.

clason commented:
When this is done, we should make sure that parsers are installed for `functionaltest-lua` as well and remove all `pending_c_parser` checks in the tests.

clason commented:
This should also include the injections for C (as well as the corresponding highlights) -- although that could be done in a followup PR together with Lua.

clason committed:
ci(tests): don't skip parsers on functionaltest

Treesitter parsers are now a mandatory part of the installation and
should be tested on all platforms. Remove `pending_c_parser` helper.

clason committed:
feat(treesitter): add viml parser and queries

clason committed:
feat(treesitter): add injections

clason commented:
I think this is pretty much ready for bikeshedding. To test this out, simply open a supported file (lua, vim, C) and do
```
:lua vim.treesitter.highlighter.new(vim.treesitter.get_parser())
```
(making sure that nvim-treesitter is not loaded). Next step is to add a `ftplugin/lua.lua` that runs this behind a `vim.g.ts_highlight_lua` guard.

Question: Do we want to include the `ERROR` nodes here? That can be quite jarring while typing. It might be better to map `@error` to `TSError` and let users link that to `Error` themselves.

bfredl commented:
> Question: Do we want to include the ERROR nodes here? That can be quite jarring while typing. It might be better to map @error to TSError and let users link that to Error themselves.

or not map `@error` at all (or use another name, like `@parse_error` maybe)

clason commented:
> or not map @error at all

yep, that's cleaner.

clason committed:
fix(treesitter): do not link @error by default

The @error capture is used for tree-sitter's ERROR node, which indicates
a parsing error -- which can be quite frequent (and jarring) while typing.

Users can still manually `hi link @error Error` in their config.

dundargoc reviewed (approved):
The cmake bit looks good, tried on both windows and linux.

bfredl commented on runtime/CMakeLists.txt:
```diff
@@ -122,6 +122,14 @@ foreach(PROG ${RUNTIME_PROGRAMS})
                  DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/nvim/runtime/${BASEDIR})
 endforeach()
 
+globrecurse_wrapper(RUNTIME_QUERIES ${CMAKE_CURRENT_SOURCE_DIR} *.scm)
+
+foreach(QUERY ${RUNTIME_QUERIES})
+  get_filename_component(BASEDIR ${QUERY} DIRECTORY)
+  install_helper(FILES ${QUERY}
+                 DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/nvim/runtime/${BASEDIR})
+endforeach()
```
are these distinct from the `RUNTIME_FILES` below, or couldn't `*.scm` just be added to its list of extensions?

clason replied:
I thought it would be useful to have these as a separate step (at least for now). But I'm sure we could combine them if you think that's better?

bfredl replied:
Well for now it is just duplicate code. it is better to separate then only if there is some extra logic (and there are note ye).

clason committed:
feat(treesitter): add vim.treesitter.start(), enable for Lua

* Add vim.treesitter.start() for starting treesitter highlighting via
  ftplugin or autocommand (can be extended later for fold, indent,
  matchpairs, ...)
* Add vim.treesitter.stop() for manually stopping treesitter
  highlighting
* Enable treesitter highlighting for Lua if
  `vim.g.ts_highlight_lua = true` is set in `init.lua`

clason commented:
With the latest commit, you can enable default treesitter highlighting for Lua (only) by adding 
```lua
vim.g.ts_highlight_lua = true
```
to your `init.lua` (needs to be run before other runtime files!)

To enable legacy syntax highlighting in parallel (e.g., if a plugin requires it), also add 
```lua
vim.g.syntax_highlight_lua = true
```
(this variable does nothing if `ts_highlight_lua` is not set).

clason commented on runtime/syntax/syntax.vim:
```diff
@@ -33,6 +33,13 @@ augroup syntaxset
   au! FileType *	0verbose exe "set syntax=" . expand("<amatch>")
 augroup END
```
I did this deliberately in order to fit the comment header but I'm fine with moving it back.

justinmk reviewed (approved):
Also adding https://github.com/vigoux/tree-sitter-vimdoc would unblock https://github.com/neovim/neovim/pull/11967

clason commented:
> Also adding https://github.com/vigoux/tree-sitter-vimdoc would unblock #11967

That is a strong argument for doing that... @vigoux?

clason commented:
I have refactored this now into a 
```lua
vim.treesitter.start(bufnr, lang, { syntax = false })
```
function that people can use for other languages they happen to have parser and highlight queries in runtimepath. I don't think we need to add another global variable to enable regex syntax highlighting in parallel; people can just put `setl syntax=on` (or `vim.treesitter.start(0, "lua", { syntax = true })`) in their own ftplugin if they need it.

clason commented:
Only two versions late, nice :)

bfredl commented:
Or it was the versions which were early :]

Liampor commented:
Does it also support c/cpp semantic highlighting with neovim native LSP?

clason commented:
No.
